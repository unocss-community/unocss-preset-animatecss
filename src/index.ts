import { Buffer } from 'node:buffer'
import type { CSSObject, Preset, Rule } from 'unocss'
import { definePreset } from 'unocss'
import postcss from 'postcss'
import { objectify, parse } from 'postcss-js'
import animatecss from './animatecss'

function toCSS(obj: any) {
  return parse(obj).toString()
}

export function presetAnimateCSS(): Preset {
  const root = postcss.parse(animatecss)
  const obj = objectify(root)

  const rootStyle = obj[':root']

  const staticRules: [string, CSSObject][] = []

  const keyframes: any[] = []

  for (const [name, body] of Object.entries(obj)) {
    if (name.startsWith('.animate__'))
      staticRules.push([name, body])
    else if (name.startsWith('@keyframes'))
      keyframes.push([name.replace('@keyframes', '').trim(), body])
  }

  /**
   * 将.a.b取b
   * @param name
   */
  const normalizeClassName = (name: string) => {
    const parts = name.split('.')
    return parts[parts.length - 1]
  }

  return definePreset({
    name: 'unocss-preset-animatecss',
    preflights: [
      {
        layer: 'animate.css',

        getCSS: () => {
          return parse({
            ':root': rootStyle,
          }).toString()
        },
      },
    ],
    layers: {
      animate: -1,
    },
    rules: [
      ...staticRules.map(([name, body]) => {
        const className = normalizeClassName(name)
        const animateName = body.animationName as string | undefined
        if (animateName) {
          const keyframe = keyframes.find(([k]) => k === animateName)
          if (keyframe) {
            const keyframeCSS = `@keyframes ${animateName} {\n${toCSS(keyframe[1])}}`
            return [new RegExp(`^${className}`),
              () => {
                return [{
                  'animation-name': animateName,
                }, keyframeCSS]
              },
            ]
          }
        }
        return [className, `${name}{
          ${toCSS(body)}
        }`]
      }) as Rule[],
    ],
  })
}

export default presetAnimateCSS
