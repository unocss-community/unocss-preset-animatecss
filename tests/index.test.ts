import { createGenerator } from 'unocss'
import { expect, test } from 'vitest'
import { presetStarter } from '../src'

test('presetStarter', async () => {
  const uno = createGenerator({
    presets: [presetStarter()],
  })
  const presets = uno.config.presets
  expect(presets).toHaveLength(1)
})
