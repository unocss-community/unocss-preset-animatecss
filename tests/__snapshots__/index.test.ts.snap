// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset-animatecss 1`] = `
"/* layer: animate.css */
:root {
  --animate-duration: 1s;
  --animate-delay: 1s;
  --animate-repeat: 1;
}
/* layer: default */
.animate__animated {
  -webkit-animation-duration: 1s;
  -webkit-animation-duration: var(--animate-duration);
  animation-duration: 1s;
  animation-duration: var(--animate-duration);
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}
.animate__animated.animate__delay-1s {
  -webkit-animation-delay: 1s;
  -webkit-animation-delay: var(--animate-delay);
  animation-delay: 1s;
  animation-delay: var(--animate-delay);
}
@keyframes backInDown {
  0% {
    -webkit-transform: translateY(-1200px) scale(0.7);
    transform: translateY(-1200px) scale(0.7);
    opacity: 0.7;
  }
  80% {
    -webkit-transform: translateY(0px) scale(0.7);
    transform: translateY(0px) scale(0.7);
    opacity: 0.7;
  }
  100% {
    -webkit-transform: scale(1);
    transform: scale(1);
    opacity: 1;
  }
}
.animate__backInDown {
  animation-name: backInDown;
}
"
`;
